### PR TITLE
Add function argument labels

### LINKED ISSUE #114: Support function argument labels

### PR DESCRIPTION
**Pull request by bearbin**
Currently labels are optional and can be mixed and matched.
Error messages should be improved.
I also didn't totally understand how to implement the ASTVisitor function, should the identifiers be visited?

### CODE REVIEW

File: Sources/AST/AST.swift [hunk 1]
```diff
@@ -743,10 +743,27 @@ public struct BinaryExpression: SourceEntity {
   }
 }
 
+/// An argument in a function call.
+public struct FunctionArgument: SourceEntity {
+  public var identifier: Identifier?
+  public var expression: Expression
+
+  public var sourceLocation: SourceLocation {
+    return expression.sourceLocation
```
Comment (DJRHails) [inline Sources/AST/AST.swift]: Consider sourceLocation returning the span of the expression

File: Sources/AST/ASTDumper.swift [hunk 2]
```diff
@@ -304,6 +304,15 @@ public class ASTDumper {
     }
   }
 
+  func dump(_ functionArgument: FunctionArgument) {
+    writeNode("FunctionArgument") {
```
Comment (DJRHails) [inline Sources/AST/ASTDumper.swift]: Add FunctionArgument to ParserTests

File: Sources/AST/Environment.swift [hunk 3]
```diff
@@ -373,15 +373,30 @@ public struct Environment {
...
+        var identifiersMatch = (candidate.parameterIdentifiers.count == argumentIdentifiers.count)
+        if identifiersMatch {
+            for (index, identifier) in candidate.parameterIdentifiers.enumerated() {
+                let matching = argumentIdentifiers[index] == nil || (argumentIdentifiers[index]!.identifierToken.kind == identifier.identifierToken.kind)
+                if !matching {
+                    identifiersMatch = matching
+                    break
+                }
+            }
+        }
```
Comment (DJRHails) [inline Sources/AST/Environment.swift]: Filter + Optional Chaining to make this a little cleaner

Comment (DJRHails) [reply Sources/AST/Environment.swift]: I need to learn how to do nicer ranges in reviews

**Comment by franklinsch** [comment]
@bearbin It should be possible to overload functions based on their labels:
```swift
func foo(a: Int) {
}

func foo(b: Int) {
}
```

I think this might take quite a lot of work to complete, and this feature is not as important as others. Maybe leave this one on the back burner for now?

**Comment by franklinsch** [comment]
Closing for now.

**Pull request closed without merging**